src/System.Text.Json/System/Text/Json/JsonReader.cs

```diff
@@ -1,383 +1,836 @@
+                case (char)JsonConstants.Quote:
+                    return ConsumeString(ref src, count);
+
+                case (char)JsonConstants.OpenBrace:
+                    StartObject();
+                    ValueType = JsonValueType.Object;
+                    return markerBytes;
+
+                case (char)JsonConstants.OpenBracket:
+                    StartArray();
+                    ValueType = JsonValueType.Array;
+                    return markerBytes;
+
+                case '0':
+                case '1':
+                case '2':
+                case '3':
+                case '4':
+                case '5':
+                case '6':
+                case '7':
+                case '8':
+                case '9':
+                    return ConsumeNumber(ref src, count, false);
+
+                case '-':
+                    int step = GetNextCharAscii(ref src, count, out char ch);
+                    if (step == 0) throw new JsonReaderException();
+                    return (ch == 'I')
+                        ? ConsumeInfinity(ref src, count, true)
+                        : ConsumeNumber(ref src, count, true);
+
+                case 'f':
+                    return ConsumeFalse(ref src, count);
+
+                case 't':
+                    return ConsumeTrue(ref src, count);
+
+                case 'n':
+                    return ConsumeNull(ref src, count);
+
+                case 'u':
+                    return ConsumeUndefined(ref src, count);
+
+                case 'N':
+                    return ConsumeNaN(ref src, count);
+
+                case 'I':
+                    return ConsumeInfinity(ref src, count, false);
+
+                case '/':
+                    // TODO: Comments?
+                    throw new NotImplementedException();
             }
 
-            if (nextByte == 'n')
-            {
-                return JsonValueType.Null;
-            }
+            return 0;
+        }
 
-            if (nextByte == '-' || (nextByte >= '0' && nextByte <= '9'))
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private int ConsumeNumber(ref byte src, int count, bool negative)
+        {
+            if (UseFastUtf8)
             {
-                return JsonValueType.Number;
+                unsafe
+                {
+                    fixed (byte* pSrc = &src)
+                    {
+                        if (!PrimitiveParser.InvariantUtf8.TryParseDecimal(pSrc, count, out decimal value, out int consumed))
+                            throw new JsonReaderException();
+
+                        // NOTE: For now, we are throwing away the resulting value. It was only used to consume all the
+                        //       numbers and discover the slice of the buffer we needed. We may want to figure out how
+                        //       to retain this value for the cases where the caller actually wants the parsed value.
```

Comment:
Instead of using primitive parser and actually parsing, is there a way to just move the cursor (and validate) but not actually get the number (which we throw away)? I would imagine the ReadInt/etc extension methods would do this instead.